### ISSUE TITLE
Make `emissary-ingress-apiext` optional

### ISSUE THREAD

**Issue by Sh1ftry** [emissary-ingress/emissary#4224]
**Please describe your use case / problem.**
I want to install Emissary-ingress without being force to use `emissary-ingress-apiext`.

**Describe the solution you'd like**
There is a way to install Emissary-ingress with `getambassador.io/v3alpha1` CRDs only.

**Additional context**
I am installing Emissary-ingress on a cluster that do not have any `getambassador.io/v2` custom resources, so I don't need support for migrating between versions.

**Comment by cindymullins-dw** [comment]
Hi @Sh1ftry, could you let us know what kind of issues this creates for you when it is included by default?

**Comment by Sh1ftry** [comment]
It's an additional component in the cluster, which needs to be monitored, regularly updated and in case of any problems it can cause disruptions in the cluster as mentioned in your [documentation](https://www.getambassador.io/docs/emissary/latest/topics/install/helm/):
```
If the emissary-ingress-apiext Deployment's Pods all stop running, you will not be able to use getambassador.io/v3alpha1 CRDs until restarting the emissary-ingress-apiext Deployment.
```
The above makes it less appealing to companies with strict regulations on what can and cannot be deployed to their clusters. This could easily be solved just by allowing users to simply disable `emissary-ingress-apiext` in case it's not needed.

**Comment by Alan01252** [comment]
Agreed I would prefer to have this opt-out if possible, it's an external dependency we don't need as we can manage the CRD install ourselves.

We've also seen weird errors where the CRDs fail with

`dry-run failed, error: conversion webhook for getambassador.io/v3alpha1, Kind=Host failed: Post "https://emissary-apiext.emissary-system.svc:443/webhooks/crd-convert?timeout=30s": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)...
`

**Comment by aaronjpitty** [comment]
Yeah +1 to this request. I experienced a similar issue to what Alan01252 is describing above. This did help (https://stackoverflow.com/a/70608787/18679153) but I shouldn't have to manually change my Firewall rules to permit apiext to run. I'd rather not have the main source for all production traffic to be reliant on a service that might not actually be needed in my particular use case.

I do think it's a fantastic addition, especially with assisting those moving from Version X to Version Y, but I think it'd be a better place to have it as an optional deployment that can be enabled / disabled in the values file (as opposed to installing it via the https://app.getambassador.io/yaml/emissary/2.2.2/emissary-crds.yaml file).

**Comment by taharah** [comment]
The webhook is configured in the CRD definition. I _think_ if you remove the block shown below from each of the CRDs, then the `emissary-ingress-apiext` resources should no longer be required. I haven't tested this yet since I ran into this issue when I was vendoring the CRD definitions to our internal environment and saw it included more resources than I expected.

```
spec:
  conversion:
    strategy: Webhook
    webhook:
      clientConfig:
        service:
          name: emissary-apiext
          namespace: emissary-system
      conversionReviewVersions:
      - v1beta1
```

**Comment by kharf** [comment]
What is the status here? I also came across this and it feels a bit like a trojan horse to have this included in a crds file. The helm chart didn't have any reference to it.
Also it does not have any resource requests, security context and readiness probe defined

**Comment by caddymofo** [comment]
We've also come across this. We fully migrated from 1.x to 2.x and also believe the api-ext deploy should be optional. Not sure if anyone else noticed this but our kube-api pods get spammed  by api-ext for failing to list/watch mapping/module/tlscontext.

```
E0701 16:39:38.022190       1 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=Mapping: conversion webhook for getambassador.io/v2, Kind=Mapping failed: no kind "Mapping" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0701 16:39:38.117232       1 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=Module: conversion webhook for getambassador.io/v2, Kind=Module failed: no kind "Module" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0701 16:39:38.721206       1 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=TLSContext: conversion webhook for getambassador.io/v2, Kind=TLSContext failed: no kind "TLSContext" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0701 16:39:39.025949       1 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=Mapping: conversion webhook for getambassador.io/v2, Kind=Mapping failed: no kind "Mapping" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0701 16:39:39.121784       1 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=Module: conversion webhook for getambassador.io/v2, Kind=Module failed: no kind "Module" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0701 16:39:39.726207       1 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=TLSContext: conversion webhook for getambassador.io/v2, Kind=TLSContext failed: no kind "TLSContext" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
```

**Comment by EdDanileyko** [comment]
Hello, I am doing a fresh installation of emissary-ingress on EKS via helm and am running into issues with the pods being able to POST to the apiext service. I don't have any necessity for v2 CRDs either and I can't figure out what is causing the problems I'm facing. It would be great to make this service optional as sometimes this issue occurs and sometimes not and I cant seem to figure out what the pattern is, but it is becoming a hindrance to my getting started.

**Comment by djsly** [comment]
@Sh1ftry I get
`* status.storedVersions[0]: Invalid value: "v1": must appear in spec.versions` when trying to remove that section 

any idea how you were able to edit the CRD ?

**Comment by phlax** [comment]
reiterating what others have said - installing a new cluster, this was by far the most problematic part of the install.

even now with everything ~working - my argocd installation continually reloads emissary due to crd mismatches - which may/not be related

**Comment by woz5999** [comment]
I was investigating why my log ingest bill had a huge, inexplicable jump. Turns out it's due to gigabytes of these logs after an emissary "upgrade" last week:

> E0701 16:39:38.022190       1 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=Mapping: conversion webhook for getambassador.io/v2, Kind=Mapping failed: no kind "Mapping" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...

The entire CRD migration path has been an absolute nightmare, time suck, and now money pit. Jeez.

**Comment by mxplusb** [comment]
> This did help (https://stackoverflow.com/a/70608787/18679153) but I shouldn't have to manually change my Firewall rules to permit apiext to run. I'd rather not have the main source for all production traffic to be reliant on a service that might not actually be needed in my particular use case.

I can confirm that adding the port to my cluster's master firewall resolved my issue, but I also agree that I shouldn't be forced to use the migration path if I have no v2 resources.

**Comment by sharkymcdongles** [comment]
So to me, it seems this api-ext setup is to allow backwards compatibility to not have to carry multiple versions of the CRD for people who use older versions of ambassador/emissary. It also seems like it may convert stuff in the other direction. We are on v2 CRDs atm, so I have patched them myself to remove all v3 and v1 things as well as the webhook. Perhaps these are useful to others: https://gist.github.com/sharkymcdongles/19d4d729ceeda694c3bc2e5737c99ea3

I only tested them with my setup, but it seems to work as far as I have tested so far albeit I just made these change 5 min ago. :D

**Comment by bchahn** [comment]
+1 for making apiext optional

**Comment by rootrahulagr** [comment]
+1

**Comment by blzhao-0** [comment]
+1

**Comment by octavian2204** [comment]
+1

**Comment by danopia** [comment]
After watching [this talk from Nov 2024](https://www.youtube.com/watch?v=aQLKdGqdrgw), I understand that emissary-ingress 4.0, when released in the nearish future (maybe before Kubecon London), will do away with the apiext webhook entirely.

One relevant slide:
![Image](https://github.com/user-attachments/assets/455c463b-223b-4259-82b0-e80ca914bb39)
